prompting/rewards/multiple_choice.py

```diff
@@ -0,0 +1,52 @@
+import time
+import torch
+import re
+from typing import List
+from prompting.rewards import BaseRewardModel, BatchRewardOutput
+from prompting.dendrite import DendriteResponseEvent
+
+
+class MultipleChoiceModel(BaseRewardModel):
+    @property
+    def name(self) -> str:
+        return "multiple_choice"
+
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.choices = ("A", "B", "C", "D")
+
+    def reward(
+        self, reference: str, response_event: DendriteResponseEvent
+    ) -> BatchRewardOutput:
+        """Compute difference scores given a completion and reference pair."""
+        rewards = []
+        timings = []
+        classes = self.choices
+        completions: List[str] = response_event.completions
+
+        for completion in completions:
+            t0 = time.time()
+            # step through the words in the completion and check if the reference is in the completion
+            # remove any punctuation too.
+            matches = [
+                word
+                for word in re.sub(r"\W", " ", completion).split()
+                if word in classes
+            ]
+
+            # Take the last match as the answer
+            if matches:
+                reward = matches[-1] == reference
```

Comment:
why do we even allow the miners to return multiple responses in the first place? Should we not try and incentivise only one clear answer scoring 0 if len(matches) == 0?

```
reward = 0 if len(matches) > 1 else matches[0] == reference
```

Reply:
The main motivation is to prevent accidental matches.  Consider the following
```
What is the name of a baby dog?

A. kitten
B. kid
*C. puppy
D. dinosaur
```
Response:
`"A baby dog is called a puppy. The answer is A."`

This should be accepted.

Reply:
if we are case insensitive, there are actually 3 matches in the response: `A, a, C`

Reply:
Yup, but this then seems like an issue with the regex being insufficient. What if:

"Answer C is correct. A baby dog is called a puppy"

The order is arbitrary as far as I understand

Reply:
What if we force to accept answer in some specific tags, eg `<ANSWER>C</ANSWER>`? Not ideal solution, but it will minimize the probability of accidental matches at least